### DISCUSSION TITLE
GPIO fast enough to sample 300 microsecond pulse?

### DISCUSSION THREAD

**Discussion by BetterAutomations** [micropython/micropython#13751] [ESP8266] [answered]
I'm using an ESP-01 (ESP8266-based) and uPy v1.22.1 to read pulses from a CAJOE v1.1 Geiger counter, but it's not triggering the interrupt routine. I can trigger the routine manually by grounding out the GPIO, but I'm wondering if the pulse is too fast for this older chip? Pulse width is about 300uS according to my oscilloscope. Do I need to upgrade to an ESP32?

**Comment by BetterAutomations** [comment]
No, I tested it with an ESP32, same result. And manual triggering works on the ESP32 as well. Confirmed with the scope that the signal drops from 3.3V to 0V for 250-300uS.

Found that [it does work with Arduino code](https://github.com/SensorsIot/Geiger-Counter-RadiationD-v1.1-CAJOE-), and I may try that next. Perhaps uPy isn't fast enough?

Sample code:
```
from micropython import const
from machine import Pin

GPIO = const(12)
counter = 0

def handler(_):
    global counter
    counter += 1
    print(repr(counter))

p = Pin(GPIO, Pin.IN, Pin.PULL_UP)
p.irq(handler=handler, trigger=Pin.IRQ_FALLING)
```

**Comment by rkompass** [comment]
Usually there is a latch in the GPIO peripheral circuitry that registers voltage changes for possible interrupts and stores the info that such a change occurred in hardware. The description in the [technical reference](https://www.espressif.com/sites/default/files/documentation/esp8266-technical_reference_en.pdf) (2.2.4)  is not very clear here.
If an interrupt of a certain change type is enabled, then the above latch is logically connected (and OR-d into) to an interrupt latch, that leads to an interrupt. When this interrupt is performed by software and how long it takes the ISR to actually register/count the interrupt and reset the above latch for enabling new interrupts depends on processor and operating system. Many other interrupt sources may have higher priorities here. The esp8266 does not shine at all in this respect.
But the hardware still should be able definitely to register a 250-300uS pulse. It should even be able to register a 250-300nS pulse.
And the software should be able to react to most pulses if they are not in very rapid succession. 
So if your intent is to simply count pulses then search for hardware issues between the Geiger tube pulse to the µC.
On the site you referenced there is a schematics involving two 555s for pulse formation. Did you use these?

PS: In your ISR code the `print(repr(counter))` definitely is not appropriate. It takes much longer than simply counting the pulse.
Doing a loop outside the ISR with a sleep of say 100 ms and doing a `print(counter)` there should be better.

**Reply by BetterAutomations** [reply, marked as answer]
Confirmed, Arduino code works. I dunno.

> On the site you referenced there is a schematics involving two 555s for pulse formation. Did you use these?

Yes, it is included in the board.

> PS: In your ISR code the print(repr(counter)) definitely is not appropriate.

It wasn't there in the first round of testing so that wasn't causing this issue, but I'll remember that for next time. I'm just going with Arduino code and done. It doesn't need to be fancy.
